pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -416,87 +432,60 @@ protected void internalCreateNonPartitionedTopic(boolean authoritative, Map<Stri
-            if (!updateLocalTopicOnly) {
-                CompletableFuture<Void> updatePartition = new CompletableFuture<>();
-                updatePartitionInOtherCluster(numPartitions, clusters).thenRun(() -> {
-                    try {
-                        namespaceResources().getPartitionedTopicResources()
-                                .updatePartitionedTopicAsync(topicName, p ->
-                            new PartitionedTopicMetadata(numPartitions)
-                        ).thenAccept(r -> updatePartition.complete(null)).exceptionally(ex -> {
-                            updatePartition.completeExceptionally(ex.getCause());
-                            return null;
-                        });
-                    } catch (Exception e) {
-                        updatePartition.completeExceptionally(e);
-                    }
-                }).exceptionally(ex -> {
-                    updatePartition.completeExceptionally(ex);
-                    return null;
-                });
-                try {
-                    updatePartition.get(DEFAULT_OPERATION_TIMEOUT_SEC, TimeUnit.SECONDS);
-                } catch (Exception e) {
-                    log.error("{} Failed to update number of partitions in zk for topic {} and partitions {}",
-                            clientAppId(), topicName, numPartitions, e);
-                    if (e.getCause() instanceof RestException) {
-                        throw (RestException) e.getCause();
+                return ret;
+            })
+            .thenCompose(__ -> {
+                final int maxPartitions = pulsar().getConfig().getMaxNumPartitionsPerPartitionedTopic();
+                if (maxPartitions > 0 && numPartitions > maxPartitions) {
+                    throw new RestException(Status.NOT_ACCEPTABLE,
+                            "Number of partitions should be less than or equal to " + maxPartitions);
+                }
+                // Only do the validation if it's the first hop.
+                if (topicName.isGlobal() && isNamespaceReplicated(topicName.getNamespaceObject())) {
+                    Set<String> clusters = getNamespaceReplicatedClusters(topicName.getNamespaceObject());
+                    if (!clusters.contains(pulsar().getConfig().getClusterName())) {
+                        log.error("[{}] local cluster is not part of replicated cluster for namespace {}",
+                                clientAppId(), topicName);
+                        throw new RestException(Status.FORBIDDEN, "Local cluster is not part of replicate cluster"
+                                + " list");
                     }
-                    throw new RestException(e);
+                    return tryCreatePartitionsAsync(numPartitions)
+                            .thenCompose(ignore -> createSubscriptions(topicName, numPartitions))
+                            .thenCompose(ignore -> {
+                                CompletableFuture<Void> ret;
```

Comment:
I think It would be clearer to use return directly.

Reply:
right. good.